File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {
            "(":")",
            "[":"]",
            "{":"}"
        }       
        stack = [] 
        for c in s:
            if c in pairs:
                stack.append(c)
            else:
                if stack and pairs[stack[-1]]==c:
                    stack.pop()
                else:
                    return False
        return len(stack)==0
        

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character,Character> pair = new HashMap<>();
        pair.put('(',')');
        pair.put('{','}');
        pair.put('[',']');

        Stack<Character> stack = new Stack<>();

        for(char c:s.toCharArray()){
            if(pair.containsKey(c)) stack.push(c);
            else{
                if(stack.empty() || pair.get(stack.pop())!=c) return false;
            }
        }

        return stack.empty();
    }
}